devices/PAA5100JE.py

```diff
@@ -0,0 +1,255 @@
+            self.spi = SPI(2, **SPIparams)
+
+        elif 'soft' in spi_port.lower():  # Works for newer versions of micropython
+            self.spi = SoftSPI(sck=Pin(sck, mode=machine.Pin.OUT, pull=machine.Pin.PULL_DOWN),
+                                       mosi=Pin(mosi, mode=machine.Pin.OUT, pull=machine.Pin.PULL_DOWN),
+                                       miso=Pin(miso, mode=machine.Pin.IN),
+                                       **SPIparams
+                                       )
+            
+        # Handle Chip Select (CS) pin
+        if spi_cs_gpio is not None:
+            self.select = Digital_output(pin=spi_cs_gpio, inverted=True)
+            self.reset = Digital_output(pin=reset, inverted=True)
+            self.select.off()  # Deselect the device by setting CS high
+            self.reset.off()
+            time.sleep(0.05)
+
+        # Reset the sensor
+        self._write(REG_POWER_UP_RESET, 0x5A)
+        time.sleep(0.02)
+        for offset in range(5):
+            self._read(REG_DATA_READY + offset)
+       
+        # Initiate registers using firmware
+        init_registers()
+
+        # Validate device ID
+        product_id = self.get_id()
+        if product_id != 0x49:
+            raise RuntimeError(f"Invalid Product ID for PAA5100: 0x{product_id:02x}")
+
+    def get_id(self):
+        """Get chip ID from PAA5100."""
+        return self._read(REG_ID, 1)
+    
+    def set_rotation(self, degrees=0):
+        """Set orientation of PAA5100 in increments of 90 degrees."""
+        if degrees == 0:
+            self.set_orientation(invert_x=True, invert_y=True, swap_xy=True)
+        elif degrees == 90:
+            self.set_orientation(invert_x=False, invert_y=True, swap_xy=False)
+        elif degrees == 180:
+            self.set_orientation(invert_x=False, invert_y=False, swap_xy=True)
+        elif degrees == 270:
+            self.set_orientation(invert_x=True, invert_y=False, swap_xy=False)
+        else:
+            raise TypeError("Degrees must be one of 0, 90, 180 or 270")
+
+    def set_orientation(self, invert_x=True, invert_y=True, swap_xy=True):
+        """Set orientation of PAA5100 manually."""
+        value = 0
+        if swap_xy:
+            value |= 0b10000000
+        if invert_y:
+            value |= 0b01000000
+        if invert_x:
+            value |= 0b00100000
+        self._write(REG_ORIENTATION, value)
+    
+    def get_motion(self):
+        buf = bytearray(12)
+        self.read_registers(REG_MOTION_BURST, buf, 12)
+        dr = buf[0]
+        x_out = to_signed_16((buf[3] << 8) | buf[2])
+        y_out = to_signed_16((buf[5] << 8) | buf[4])
+        quality = buf[6]
+        shutter_upper = buf[10]
+        if (dr & 0b10000000) and not ((quality < 0x19) and (shutter_upper == 0x1f)):
+            return x_out, y_out
+        else:
+            x_out, y_out = 0, 0
+        time.sleep_ms(1)
+
+        return x_out, y_out
+    
+    def _write(self, register, value):

```

Comment:
being pedantic: please use type annotations and docstring for the functions so it's easier to develop the later on